contracts/DssVestTopUp2.0.sol

```diff
@@ -0,0 +1,352 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import "@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol";
+import "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
+import "./interfaces/IUpkeepRefunder.sol";
+
+interface DssVestLike {
+    function vest(uint256 _id) external;
+
+    function unpaid(uint256 _id) external view returns (uint256 amt);
+}
+
+interface DaiJoinLike {
+    function join(address usr, uint256 wad) external;
+}
+
+interface KeeperRegistryLike {
+    struct UpkeepInfo {
+        address target;
+        uint32 executeGas;
+        bytes checkData;
+        uint96 balance;
+        address admin;
+        uint64 maxValidBlocknumber;
+        uint32 lastPerformBlockNumber;
+        uint96 amountSpent;
+        bool paused;
+        bytes offchainConfig;
+    }
+
+    function getUpkeep(
+        uint256 id
+    ) external view returns (UpkeepInfo memory upkeepInfo);
+
+    function addFunds(uint256 id, uint96 amount) external;
+}
+
+/**
+ * @title DssVestTopUp
+ * @notice Replenishes upkeep balance on demand
+ * @dev Withdraws vested tokens or uses transferred tokens from Maker protocol and
+ * funds an upkeep after swapping the payment tokens for LINK
+ */
+contract DssVestTopUp20 is IUpkeepRefunder, Ownable {
```

Comment:
What are the steps to avoid duplicating the contract?

Note: In general, if you upgrade something, all dependencies must be updated - something that need to be considered before making such major decisions.

Reply:
We need to update DssVestTopUp unit tests. I thought that this way if we can have two versions one with the old KeeperRegistryLike and one with the newer KeeperRegistryLike we can "patch" it.

Reply:
What exactly needs to be updated in the unit tests?

Also, what about the deploy and staging (including utils) scripts? Do they need to be updated?

If there are too many changes it might be better to upgrade to v2 in a separate PR.

Reply:
If we are going to use KeeperRegistry 2.0 we have to create an upkeep in the staging script and in tests because of this [line](https://github.com/hackbg/chainlink-makerdao-keepers/pull/63/files#diff-eab3ab5496630b902fa8f3fedaa7a2d79fedf0457a6faa34e4bc361bac1ec967R171) :

`uint96 balance = keeperRegistry.getUpkeep(upkeepId).balance;`  

If there isn't an upkeep active registered upkeep there isn't `.balance` for that upkeep and the contract reverts. We might need to create a separate PR for that upgrade to V2 as this PR is already "heavy" like you mentioned.